### ISSUE TITLE
Improve error message for: Error: Value is null at ensureNotNull (lightweight-charts.development.mjs:205:15)

### ISSUE THREAD

**Issue by 0x33dm** [tradingview/lightweight-charts#1668]
**Lightweight Charts™ Version:** 

4.2.0 🚭 

**Steps/code to reproduce:**

I have a React Hook that updates lines, candles and Area chart based on different states of my app.

It generally works but sometimes i get that error

> Error: Value is null at ensureNotNull (lightweight-charts.development.mjs:205:15)

When I look through the stack trace, the further I can go is a `requestAnimationFrame`; making it very hard to figure out where the null value is coming from.

Ideally, the error message would have a label indicating what kind of value is null, making it easier to debug and figure out where the null value is coming from.

When this happens, the entire chart doesn't render, which is a big issue. It also makes me wonder if some sort of try/catch should exist in the library so that other parts of the chart would still render and not get affected if only a LINE or an AREA chart has a null value.

I'm assuming it's very unlikely the problem is with the candles as the candles actually don't change so it would be really helpful if at least the candles would render.

Still, there is a chance a candle that comes with `null` from a third-party API from time to time could have a null value (I'm still figuring it out).

Unfortunately, this error occurs very rarely, making it difficult to reproduce and determine the source of the problem.

Could we add a string to this error pointing out which type of value or series the problem is?

**Comment by SlicedSilver** [comment]
`ensureNotNull` is typically used for internal data, or references to dom elements. I would expect that the data is validated when provided when using the development version of the library and you would then get a better error message.

Is it possible that you are mutating the data after it is provided to the library?

As a check I would suggest cloning any data before you provide it to the library (like with a `JSON.parse(JSON.stringify(data))`) so that some other code elsewhere in your react application isn't modifying it.

If you can get a stack trace for the error then we might be able to assist with narrowing down which part is breaking.

**Comment by 0x33dm** [comment]
@SlicedSilver, thanks for the help, I'll be double-checking if i'm mutating something.

I think on of the most delicate part of it is that i hold a "ref" to the DOM / to the chart, like this, so i wonder if there might be a race condition somewhere, even tough i'm checking for those refs everywhere...

```ts
    if (!chartContainerRef.current || !candles.length) {
      return
    }

    // don't create chart twice
    if (chartRef.current) {
      return
    }

    chartRef.current = createChart(
      chartContainerRef.current,
      getInitialChartOptions(chartColors)
    )
```

Here is a screenshot of the error on the console:

![image](https://github.com/user-attachments/assets/c81ddeca-5c2c-4859-a5eb-dfc2050b1609)

![image](https://github.com/user-attachments/assets/ddd70125-64fc-4cf4-8a32-b90987f7817b)

**Comment by 0x33dm** [comment]
Here is another error that for some reason gave me the option to "show more frames":

![image](https://github.com/user-attachments/assets/b54bc102-2923-494d-a386-3f9e9f273555)

**Comment by SlicedSilver** [comment]
Based on the error message, this is the line of source code which is throwing the error:
https://github.com/tradingview/lightweight-charts/blob/fcc1fe338dea7e386effa891bf20e09030707819/src/views/pane/crosshair-marks-pane-view.ts#L88

It is fetching the first value for the series and then checking that it is not null.

It fetches the value via https://github.com/tradingview/lightweight-charts/blob/fcc1fe338dea7e386effa891bf20e09030707819/src/model/series.ts#L352

which gets it from:
https://github.com/tradingview/lightweight-charts/blob/fcc1fe338dea7e386effa891bf20e09030707819/src/model/series.ts#L364-L372

So one possible issue is that the chart is momentarily becoming very small or detached from the DOM and thus the time scale width is 0 and there is no visible range. This might be caused by React performing some DOM operations, and thus I would suggest only rendering the container and chart once and then using the existing API to make any changes to the chart.

The other possibility is that the data contains missing or deleted values (perhaps via data mutation).

Perhaps you could add a breakpoint on that line of code that can return null and see which one is causing the issue.

**Comment by 0x33dm** [comment]
> So one possible issue is that the chart is momentarily becoming very small or detached from the DOM and thus the time scale width is 0 and there is no visible range. This might be caused by React performing some DOM operations, and thus I would suggest only rendering the container and chart once and then using the existing API to make any changes to the chart.

Currently, my setup avoids creating the chart multiple times, the code is +/- this:

```ts
function setupChart() {
    // don't create chart twice
    if (chartRef.current) {
      return
    }

    chartRef.current = createChart(
      chartContainerRef.current,
      getInitialChartOptions(chartColors)
    )
}
```

So, the chart is created only once.


I thought about adding try/catch around every place i update the data but i believe this would probably not work because the error happens with a `requestAnimationFrame` so i'm assuming the try/catch wouldn't catch that...?

**Comment by 0x33dm** [comment]
Another thing i'm suspecting is that perhaps i'm calling `setData` on the candlesSeries a few times in a row ( as in: not debounced ) due to it depending on multiple hooks.

i wonder if that could be the reason

**Comment by 0x33dm** [comment]
> Another thing i'm suspecting is that perhaps i'm calling `setData` on the candlesSeries a few times in a row ( as in: not debounced ) due to it depending on multiple hooks.
> 
> i wonder if that could be the reason

I was always "replacing" the candles this way:

```ts
    mainSeriesRef.current?.setData(candles)
```

Now I'm trying this way:

```ts
    mainSeriesRef.current?.setData([])
    mainSeriesRef.current?.setData(candles)
```

perhaps the "mutated" data could be the "old candles" that got replaced... ?

maybe there is a better way of "replacing" the candles?

**Comment by SlicedSilver** [comment]
I don't know the specifics of your implementation but maybe we could build a reproduction by:
- adding a console.log each time you set the candles data (wrapped in JSON.stringify).
- If you manage to get the issue to occur then you will be able to see the exact data updates which caused the issue
- and then create a small example which sets that same data.

If the `candles` array is being controlled by some other piece of code (perhaps react) then it could be doing some mutations. The library itself will do some mutations on the data it receives so it expects that nothing else will be changing the values. 

Could you try:
```
mainSeriesRef.current?.setData(JSON.parse(JSON.stringify(candles)))
```
or [structuredClone](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone).

**Comment by 0x33dm** [comment]
After a lot of debugging it came down to updating the candle data. I was cloning the candles before by doing:

```ts
const newCandles = [...candles]
```

So somehow, I assumed JSON clone wouldn't make much of a difference.

In the end setting candles to `[]` before setting the new candles is doing the job ( also cheaper than JSON / structuredClone ).

I haven't seen any more crashes since. I wonder if that should be the default behaviour of the library. If so, every time you setData for candles, it should first erase the first one, like I'm doing here manually, and then set the new one. 

Does keeping the old array supports any feature library or would always erasing also works?

I assume `setData` will always overwrite the current data or will it merge it somehow?

Thank you!

**Comment by SlicedSilver** [comment]
The library isn't responsible for the cloning of the data, and it will use the values as provided.

With `setData` we aren't merging it with an existing array or keeping the old array for anything. So there isn't really anything the library can do extra here besides making it's own copy of the data as soon as it is provided to the library.

The reason why we don't do this is performance and memory usage. Doing an extra clone of data takes a bit of CPU time, and the data is then duplicated in the memory. For most use cases this overhead isn't required, and for the few use-cases where it is required (like here with a React hook) then the responsibility for providing data which won't be mutated can be handled externally to the library.

TLDR; we don't do internal cloning of the data provided for performance reasons.

---

```js
const candles = [{ time: 1, value: 10 }];

// mutate the data (either externally by your code, react, or internally by the library)
candles[0].value = 20;

const newCandles = [...candles, { time: 2, value: 10 }];
console.log(newCandles); // → [{ time: 1, value: 20 }, { time: 2, value: 10 }]
```
👆 new candles still contains the modified values

Clearing the candles array before creating the newCandles array will have a different end result:
```js
let candles = [{ time: 1, value: 10 }];
candles[0].value = 20;

candles = [];
const newCandles = [...candles, { time: 2, value: 10 }];
console.log(newCandles); // → [{ time: 2, value: 10 }]
```
👆 new candles doesn't contain any of the old candles

**Comment by 0x33dm** [comment]
> The library isn't responsible for the cloning of the data, and it will use the values as provided.
> 
> With `setData` we aren't merging it with an existing array or keeping the old array for anything. So there isn't really anything the library can do extra here besides making it's own copy of the data as soon as it is provided to the library.
> 
> The reason why we don't do this is performance and memory usage. Doing an extra clone of data takes a bit of CPU time, and the data is then duplicated in the memory. For most use cases this overhead isn't required, and for the few use-cases where it is required (like here with a React hook) then the responsibility for providing data which won't be mutated can be handled externally to the library.
> 
> TLDR; we don't do internal cloning of the data provided for performance reasons.
> 
> ```js
> const candles = [{ time: 1, value: 10 }];
> 
> // mutate the data (either externally by your code, react, or internally by the library)
> candles[0].value = 20;
> 
> const newCandles = [...candles, { time: 2, value: 10 }];
> console.log(newCandles); // → [{ time: 1, value: 20 }, { time: 2, value: 10 }]
> ```
> 
> 👆 new candles still contains the modified values
> 
> Clearing the candles array before creating the newCandles array will have a different end result:
> 
> ```js
> let candles = [{ time: 1, value: 10 }];
> candles[0].value = 20;
> 
> candles = [];
> const newCandles = [...candles, { time: 2, value: 10 }];
> console.log(newCandles); // → [{ time: 2, value: 10 }]
> ```
> 
> 👆 new candles doesn't contain any of the old candles

Sure thing, i understand all of that.

What i suggested tough is when we do `setData` for the candles the library should internally "erase" the previous array.

In my case i do `setData([])` to avoid that problem which i guessed the library could also do internally to avoid the problem i was having.

TLDR:

Hook with this triggers the problem:
```ts
    mainSeriesRef.current?.setData(candles)
```

Hook with this, doesn't trigger the problem:
```ts
mainSeriesRef.current?.setData([])
mainSeriesRef.current?.setData(candles)
```

Conclusion: maybe internally when new candles are set the previous candles could be erased the same way i did.

**Comment by SlicedSilver** [comment]
I'll adjust my statement slightly to say that we [do have access to the old data (aka current data)](https://github.com/tradingview/lightweight-charts/blob/fcc1fe338dea7e386effa891bf20e09030707819/src/model/data-layer.ts#L152-L158) when the `setData` is providing the new data so we could technically do the erasing on your behalf (after the comparison checks used for performance optimisations). Essentially calling `setData` twice (once to clear and once to set). 

I'm not convinced yet that we would want to enforce this behaviour internally when it could be done (optionally) externally to the library. Seems like a good thing to document further (perhaps on the `setData` description).

It's a good edge case to be aware of for React users and I'm sure your discovery will help some other people when searching for similar bug reports. Thanks for investigating and finding a work around solution for the problem.

**Comment by 0x33dm** [comment]
> I'll adjust my statement slightly to say that we [do have access to the old data (aka current data)](https://github.com/tradingview/lightweight-charts/blob/fcc1fe338dea7e386effa891bf20e09030707819/src/model/data-layer.ts#L152-L158) when the `setData` is providing the new data so we could technically do the erasing on your behalf (after the comparison checks used for performance optimisations). Essentially calling `setData` twice (once to clear and once to set).
> 
> I'm not convinced yet that we would want to enforce this behaviour internally when it could be done (optionally) externally to the library. Seems like a good thing to document further (perhaps on the `setData` description).
> 
> It's a good edge case to be aware of for React users and I'm sure your discovery will help some other people when searching for similar bug reports. Thanks for investigating and finding a work around solution for the problem.

Do you think is there a case where the performance would be better if we do the structuredClone or JSON clone instead of erasing it first?

For instance, the library would compare every data point and only redraw the ones that changed?

If that's not the case then might be worth always erasing the previous one.

Unfortunately i don't understand enough of the library to have an opinion here, anyway i'll close the issue as i found this solution.

thank you

**Comment by SlicedSilver** [comment]
For the best performance, try use `update` instead of `setData` if possible.
https://tradingview.github.io/lightweight-charts/docs#updating-the-data-in-a-series

There are some cases where you would need to use `setData` (such as adjusting previous values) but if you are only needing to update the last point or add more points to the existing dataset then `update` is a lot faster then using `setData`

**Comment by 0x33dm** [comment]
> For the best performance, try use `update` instead of `setData` if possible. https://tradingview.github.io/lightweight-charts/docs#updating-the-data-in-a-series
> 
> There are some cases where you would need to use `setData` (such as adjusting previous values) but if you are only needing to update the last point or add more points to the existing dataset then `update` is a lot faster then using `setData`

We just call the API for the last Y candles every X seconds and set the data.

Ultimately, I think we don't know if we're updating the last candle or adding a new one ( for instance, if a candle has just closed ).

I guess the update would expect us to update just the "diff" between last candles and the new candles?

ps: thanks a lot for your help and support.

**Comment by SlicedSilver** [comment]
If you were to keep track of the last timestamp of data that you've provided to the library then you could do something like this (types might be wrong since I'm doing this from memory):
```ts
function setNewData(newData: LineData[], currentLastTimestamp: Time, series: ISeriesApi<'Line'>): Time {
   // assuming that your 'Time' is a timestamp number, then
  const updateData = newData.filter((d: LineData) => d.time >= currentLastTimestamp);
  updateData.forEach((d: LineData) => {
    /*
        you can only use a single data point for update,
        so we can do this in a loop if we have potentially
        more than one point
    */
    series.update(d);
  });
  return updateData[updateData.length - 1].time; // new current last timestamp
}
```

**Comment by MarvinMiles** [comment]
Had same error with no clue where to start digging in to resolve issue.

> TLDR:
> 
> Hook with this triggers the problem:
> 
> ```ts
>     mainSeriesRef.current?.setData(candles)
> ```
> 
> Hook with this, doesn't trigger the problem:
> 
> ```ts
> mainSeriesRef.current?.setData([])
> mainSeriesRef.current?.setData(candles)
> ```
> 
> Conclusion: maybe internally when new candles are set the previous candles could be erased the same way i did.

This was helpful. Thanks

**Comment by 0x33dm** [comment]
> Had same error with no clue where to start digging in to resolve issue.
> 
> > TLDR:
> > Hook with this triggers the problem:
> > ```ts
> > mainSeriesRef.current?.setData(candles)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hook with this, doesn't trigger the problem:
> > ```ts
> > mainSeriesRef.current?.setData([])
> > mainSeriesRef.current?.setData(candles)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Conclusion: maybe internally when new candles are set the previous candles could be erased the same way i did.
> 
> This was helpful. Thanks

glad it worked for you. it took me a long time to figure this out and now it's solid and i'm not having crashes anymore.

It was a big issue for my application for several days.

I think the library should be doing this `setData([])` by itself but perhaps I just don't understand what the downside would be.
